master/internal/provisioner/aws_spot_legacy.go

```diff
@@ -0,0 +1,216 @@
+
+	// List spot requests with the old format
+	activeSpotReqs, err := c.legacyListActiveSpotInstanceRequests(ctx)
+	if err != nil {
+		loggerSpotLegacy.
+			WithError(err).
+			Errorf("cannot list active legacy spot requests")
+		return
+	}
+	if activeSpotReqs.numReqs() == 0 {
+		loggerSpotLegacy.Debugf("no active legacy spot requests to clean up")
+		return
+	}
+	// Delete spot requests with old format
+	loggerSpotLegacy.Infof(
+		"terminating %d active legacy spot requests: %s",
+		activeSpotReqs.numReqs(),
+		strings.Join(activeSpotReqs.idsAsList(), ","))
+
+	_, err = c.terminateSpotInstanceRequests(ctx, activeSpotReqs.idsAsListOfPointers(), false)
+	if err != nil {
+		loggerSpotLegacy.
+			WithError(err).
+			Errorf("unable to terminate active legacy spot requests")
+	}
+
+	// Delete spot instances associated with the active requests
+	instancesToTerminate := newSetOfStrings()
+	for _, req := range activeSpotReqs.iter() {
+		if req.InstanceID != nil {
+			instancesToTerminate.add(*req.InstanceID)
+		}
+	}
+	if instancesToTerminate.length() == 0 {
+		loggerSpotLegacy.Debugf("no instances associated with active legacy spot requests to terminate")
+		return
+	}
+
+	loggerSpotLegacy.Infof(
+		"terminating %d legacy spot instances associated with active legacy spot requests: %s",
+		instancesToTerminate.length(),
+		strings.Join(instancesToTerminate.asList(), ","))
+
+	_, err = c.terminateInstances(instancesToTerminate.asListOfPointers())
+	if err != nil {
+		loggerSpotLegacy.
+			WithError(err).
+			Errorf("unable to terminate instances associated with active legacy spot requests")
+	}
+}
+
+func (c *awsCluster) legacyCleanupCanceledButInstanceRunningSpot(ctx *actor.Context) {
+	loggerSpotLegacy := ctx.Log().WithField("codepath", "spotLegacy")
+
+	loggerSpotLegacy.Debugf("listing CanceledButInstanceRunning requests")
+	canceledButInstanceRunningSpotReqs, err := c.legacyListCanceledButInstanceRunningSpotRequests(ctx)
+	if err != nil {
+		loggerSpotLegacy.
+			WithError(err).
+			Debugf("unable to list CanceledButInstanceRunning legacy spot requests")
+		return
+	}
+	// Delete any instances associated with canceledButInstanceRunning spot requests
+	if canceledButInstanceRunningSpotReqs.numReqs() == 0 {
+		loggerSpotLegacy.Debugf("no CanceledButInstanceRunning legacy spot requests to clean up")
+		return
+	}
+
+	loggerSpotLegacy.
+		Infof(
+			"terminating %d legacy spot instances where requests are CanceledButInstanceRunning: %s",
+			canceledButInstanceRunningSpotReqs.numReqs(),
+			strings.Join(canceledButInstanceRunningSpotReqs.idsAsList(), ","),
+		)
+
+	_, err = c.terminateInstances(canceledButInstanceRunningSpotReqs.instanceIds())
+	if err != nil {
+		loggerSpotLegacy.
+			WithError(err).
+			Debugf("cannot terminate EC2 instances associated with canceled spot requests")
+	}
+}
+
+func isLegacy(request *ec2.SpotInstanceRequest) bool {
+	for _, tag := range request.Tags {
+		if *tag.Key == "determined-resource-pool" {
```

Comment:
I feel like this deref is probably always safe (unless maybe the response is malformed) but i'd feel better if we checked it first since this would crash the provisioner.

Reply:
fixed